pytorch_lightning/loggers/mlflow.py

```diff
@@ -266,3 +291,70 @@ def version(self) -> str:
             The run id.
         """
         return self.run_id
+
+    def after_save_checkpoint(self, checkpoint_callback: "ReferenceType[ModelCheckpoint]") -> None:
+        # log checkpoints as artifacts
+        if self._log_model == "all" or self._log_model is True and checkpoint_callback.save_top_k == -1:
+            self._scan_and_log_checkpoints(checkpoint_callback)
+        elif self._log_model is True:
+            self._checkpoint_callback = checkpoint_callback
+
+    def _scan_and_log_checkpoints(self, checkpoint_callback: "ReferenceType[ModelCheckpoint]") -> None:
+        # get checkpoints to be saved with associated score
+        checkpoints = {
```

Comment:
Can this function can converted into a utility function and shared between loggers ?

Reply:
Yes, I was wondering the same when writing it, but I was not sure where to put it. An idea is to split the function in:
<details>
  <summary markdown="span"

 `_scan_checkpoints(self, checkpoint_callback: "ReferenceType[ModelCheckpoint]") -> List[ModelCheckpoint]` 

</summary>

```
_scan_checkpoints(self, checkpoint_callback: "ReferenceType[ModelCheckpoint]") -> List[ModelCheckpoint]:
    # get checkpoints to be saved with associated score
    checkpoints = {
        checkpoint_callback.last_model_path: checkpoint_callback.current_score,
        checkpoint_callback.best_model_path: checkpoint_callback.best_model_score,
        **checkpoint_callback.best_k_models,
    }
    checkpoints = sorted((Path(p).stat().st_mtime, p, s) for p, s in checkpoints.items() if Path(p).is_file())
    checkpoints = [
        c for c in checkpoints if c[1] not in self._logged_model_time.keys() or self._logged_model_time[c[1]] < c[0]
    ]
    return checkpoints
```

</details>

which is generic and can be probably put inside `Base.py`, and:

<details>
  <summary markdown="span"

` _log_checkpoints(self, checkpoints: List[ModelCheckpoint]) -> None:`  for `MLFlowLogger.py`

 </summary>

```
_log_checkpoints(self, checkpoints: List[ModelCheckpoint]) -> None:
    # log iteratively all new checkpoints
    for t, p, s in checkpoints:
    metadata = {
        # Ensure .item() is called to store Tensor contents
        "score": s.item() if isinstance(s, torch.Tensor) else s,
        "original_filename": Path(p).name,
        "ModelCheckpoint": {
            k: getattr(checkpoint_callback, k)
            for k in [
                "monitor",
                "mode",
                "save_last",
                "save_top_k",
                "save_weights_only",
                "_every_n_train_steps",
                "_every_n_val_epochs",
            ]
            # ensure it does not break if `ModelCheckpoint` args change
            if hasattr(checkpoint_callback, k)
        },
    }
    aliases = ["latest", "best"] if p == checkpoint_callback.best_model_path else ["latest"]
    
    # Artifact path on mlflow
    artifact_path = f"model/checkpoints/{Path(p).stem}"
    
    # Log the checkpoint
    self.experiment.log_artifact(self._run_id, p, f"{artifact_path}")
    
    # Create a temporary directory to log on mlflow
    with tempfile.TemporaryDirectory() as tmp_dir:
        # Log the metadata
        with open(f"{tmp_dir}/metadata.yml", "w") as tmp_file_metadata:
            yaml.dump(metadata, tmp_file_metadata, default_flow_style=False)
    
        # Log the aliases
        with open(f"{tmp_dir}/aliases.txt", "w") as tmp_file_aliases:
            tmp_file_aliases.write(str(aliases))
    
        # Log the metadata and aliases
        self.experiment.log_artifacts(self._run_id, tmp_dir, artifact_path)
    
    # remember logged models - timestamp needed in case filename didn't change (lastkckpt or custom name)
    self._logged_model_time[p] = t
```

</details>

<details>
  <summary markdown="span"

` _log_checkpoints(self, checkpoints: List[ModelCheckpoint]) -> None:`  for `WandbLogger.py`

 </summary>

```
_log_checkpoints(self, checkpoints: List[ModelCheckpoint]) -> None:
    # log iteratively all new checkpoints
    for t, p, s in checkpoints:
    metadata = (
        {
            "score": s,
            "original_filename": Path(p).name,
            "ModelCheckpoint": {
                k: getattr(checkpoint_callback, k)
                for k in [
                    "monitor",
                    "mode",
                    "save_last",
                    "save_top_k",
                    "save_weights_only",
                    "_every_n_train_steps",
                    "_every_n_val_epochs",
                ]
                # ensure it does not break if `ModelCheckpoint` args change
                if hasattr(checkpoint_callback, k)
            },
        }
        if _WANDB_GREATER_EQUAL_0_10_22
        else None
    )
    artifact = wandb.Artifact(name=f"model-{self.experiment.id}", type="model", metadata=metadata)
    artifact.add_file(p, name="model.ckpt")
    aliases = ["latest", "best"] if p == checkpoint_callback.best_model_path else ["latest"]
    self.experiment.log_artifact(artifact, aliases=aliases)
    # remember logged models - timestamp needed in case filename didn't change (lastkckpt or custom name)
    self._logged_model_time[p] = t
```

</details>

Reply:
Added this in the last commits